File: 1016-subarray-sums-divisible-by-k/subarray-sums-divisible-by-k.c
#include <stdio.h>
#include <stdlib.h>

int subarraysDivByK(int* nums, int numsSize, int k) {
    int* dp = calloc(k, sizeof(int));
    dp[0] = 1;

    int res = 0;
    int x = 0;

    for (int i = 0; i < numsSize; i++) {
        x += nums[i];
        int y = ((x % k) + k) % k;
        res += dp[y];
        dp[y]++;
    }

    free(dp);
    return res;
}

File: 1016-subarray-sums-divisible-by-k/subarray-sums-divisible-by-k.py
class Solution(object):
    def subarraysDivByK(self, nums, k):
        dp = defaultdict(int)
        dp[0] = 1

        res = 0
        x = 0

        for i in nums:
            x += i
            y = ((x % k) + k) % k
            res += dp[y]
            dp[y] += 1
        
        return res
